src/csharp/Grpc.Core/Internal/AsyncCall.cs

```diff
@@ -510,6 +510,7 @@ private void HandleFinished(bool success, ClientSideStatus receivedStatus)
 
             lock (myLock)
             {
+                readingDone = true;
```

Comment:
The problem with doing just this is (as indicated by tests):
1. in ReadMessageInternalAsync() we are using readingDone and relying on streamingReadTcs being populated (once the stream has finished, we keep returning the result of the last read).
2. if there's a read already in flight (read has started but the op hasn't finished yet), we have an issue - we would need believe that that read is going to end up with "null" message and the call would be disposed on receiving that, but technically, that's not guaranteed and we could receive a message (and in that case we would keep delivering the same message over and over).

So, I think these can be fixed, but you probably need to do a few more adjustments.

Suggestions worth trying:
1. if   streamingReadTcs == null (no read in flight), assign a new TCS and populate it with .SetResult(default(TResponse)).
2. if a read is in flight, let it finish - there's a very good chance it's going to be a null message and if so problem solved) - if the received message is non-null, things become pretty tricky because theoretically you could still receive another message from C core (i'll have to check if that's the case actually), so you should not cut off the stream (actually could a problem in case 1 as well).